Ticker free() - requirements, pseudo code, tests, implementation

mprse opened the pull request:
### Description

This patch fixes ticker `ticker_free()` issue. 

> The ticker HAL API has been well defined (defined/undefined behaviour, potential bugs) in the ticker header files the tests which verifies each requirement have been also created. We had all done.
> But meanwhile lp/us_ticker_free() functions have been added: https://github.com/ARMmbed/mbed-os/pull/5856
> 
> This PR added lp/us_ticker_free() functions to the ticker HAL header files and implementation for only some targets, but did not specify the requirements for these functions (defined/undefined behaviour, potential bugs) as this has been done for other functions.
> As a result we have now ready requirements/tests and function in HAL API which is not verified.


@c1728p9 Can you please take a look at the requirements, tests and example implementation?

### TODO

- [x] Add requirements and pseudo code for HAL ticker_free() function.
- [X] Add ticker_free() function to the ticker interface.
- [X] Add ticker_free() functional tests.
- [X] Add example implementation of `ticker_free()` for one target for verification (`NRF5x` family boards).
- [X] Add/update `ticker_free()` implementation for the remaining boards.

### Pull request type

    [] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

mprse committed:
Add ticker_free() function to the ticker interface.

mprse committed:
tests-mbed_hal-common_tickers: add ticker_free() functional tests.

mprse committed:
Add requirements and pseudo code for HAL ticker_free() function.

NirSonnenschein commented:
build failures in CI:
 Error: L6218E: Undefined symbol lp_ticker_free (referred from BUILD/tests/K64F/ARM/hal/mbed_lp_ticker_api.o 
please take a look

mprse commented:
> build failures in CI:
> Error: L6218E: Undefined symbol lp_ticker_free (referred from BUILD/tests/K64F/ARM/hal/mbed_lp_ticker_api.o
> please take a look

Yes, like I mentioned in the description. Implementation of `ticker_free` needs to be added for some targets, for others needs to be adapted to the requirements. But first we need to agree the requirements.

mprse commented:
@c1728p9 @0xc0170 I'm not sure if requirements for '`ticker_free()` function meets expectation. Can you please take a look. 

Should we use `ticker_free()` for example before going to deep-sleep in order to disable us ticker (like on `NRF` boards)? In such case the counter value should persist and after wake up us ticker should be re-initialised by means of `ticker_init` and continue counting.
Or maybe this function shell simply turn of the ticker, ticker interrupts and clock gate.

c1728p9 commented on hal/us_ticker_api.h:
```diff
@@ -73,6 +73,9 @@ extern "C" {
  * Verified by ::ticker_fire_now_test
  * * The ticker operations ticker_read, ticker_clear_interrupt, ticker_set_interrupt and ticker_fire_interrupt
  * take less than 20us to complete - Verified by ::ticker_speed_test
+ * * The function ticker_free stops the ticker from counting and disables the ticker interrupt -
```
It shouldn't be a requirement that the ticker stops counting, since some low power ticker implementations are backed by an RTC which may keep counting.

c1728p9 reviewed (commented):
Looks good aside from the requirement that the ticker stops counting.

mprse committed:
Add implementation of ticker_free() function to CI boards.

This PR provides implementation of ticker_free() function for the following boards:
ARCH_PRO
EV_COG_AD3029LZ
EV_COG_AD4050LZ
K22F
K64F
K82F
KW24D
KW41Z
LPC546XX
NRF51_DK
NRF52_DK
NUCLEO_F207ZG
NUCLEO_F401RE
NUCLEO_F429ZI
NUCLEO_F746ZG
REALTEK_RTL8195AM

mprse commented:
Fixed after review and added implementation for CI boards.

mprse committed:
TESTS-MBED_DRIVERS-TIMER: include ticker_free() function in the ticker interface.

bcostm commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -231,7 +231,8 @@ void lp_ticker_clear_interrupt(void)
 void lp_ticker_free(void)
 {
-
+    LptimHandle.Instance = LPTIM1;
```
This line is useless as the LptimHandle is not used.

bcostm commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -278,12 +279,12 @@ void lp_ticker_disable_interrupt(void)
 void lp_ticker_free(void)
 {
-
+    //NVIC_DisableIRQ(RTC_WKUP_IRQn);
```
Remove this line ?

bcostm commented on targets/TARGET_STM/lp_ticker.c:
```diff
@@ -231,7 +231,8 @@ void lp_ticker_clear_interrupt(void)
 {
-
+    LptimHandle.Instance = LPTIM1;
+    NVIC_DisableIRQ(LPTIM1_IRQn);
```
Why not call the lp_ticker_disable_interrupt() instead ?

bcostm commented on targets/TARGET_STM/us_ticker.c:
```diff
@@ -259,6 +259,6 @@ void restore_timer_ctx(void)
 void us_ticker_free(void)
 {
-
+    __HAL_TIM_DISABLE_IT(&TimMasterHandle, TIM_IT_CC1);
```
Why not call the us_ticker_disable_interrupt() instead ?

bcostm reviewed (commented):
LGTM. I left few comments.

mprse commented:
@bcostm Thanks for the review.
I followed your hints and changed approach a little. So according to the requirements `ticker_free()` at least has to disable ticker interrupt and would be good to disable also the hardware counter  (at least for non RTC based tickers).

cmonr commented:
@c1728p9 @bcostm Mind taking another look? Looks like y'alls comments have been addressed.

0xc0170 commented:
/morph build

0xc0170 commented:
@mprse I believe test will fail as some targets need free implemented

mprse commented:
> @mprse I believe test will fail as some targets need free implemented

Hopefully I manage to add ticker free for all targets. For CI targets ticker free is implemented to be consistent with requirements (I couldn't test all CI boards since I do not have all of them). For other boards I just added empty stub of ticker free function.

mprse commented:
```
./TESTS/mbedmicro-rtos-mbed/systimer/main.cpp
        [DEBUG] Return: 1
        [DEBUG] Output: ./TESTS/mbedmicro-rtos-mbed/systimer/main.cpp:119:1: sorry, unimplemented: non-trivial designated initializers not supported
        [DEBUG] Output:  };
        [DEBUG] Output:  ^
```

Test for systimer initialises ticker interface with stubs. This needs to be updated and ticker free function must be included.

mprse committed:
TESTS-MBEDMICRO-RTOS-MBED-SYSTIMER:  include ticker_free() function in the ticker interface.

mprse commented:
Fixed `TESTS-MBEDMICRO-RTOS-MBED-SYSTIMER`. Can we try tun morph again?

0xc0170 commented:
/morph build

mprse committed:
Add lp/us ticker_free() functions stub.

This patch adds only empty stubs of `us_ticker_free()` and `lp_ticker_free()` for all boards where these functions are not implemented.

mprse commented:
> Build : FAILURE
> Build number : 2663
> Build artifacts/logs : http://mbed-os.s3-website-eu-west-1.amazonaws.com/?prefix=builds/7508/

`FVP_MPS2`: `us_ticker_free()` has been re-implemented by mistake - fixed.
`TMPM3H6`: target has been added meanwhile without implementation of `us_ticker_free()`. Rebased this PR and added missing implementation.

Can we try again with morph?

0xc0170 commented:
/morph build

mprse commented:
> Test : FAILURE
> Build number : 2407
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/7508/2407
> 

For the following boards `ticker free` tests have failed:

- K22F - bug, `PIT_StartTimer` was called by `ticker free()` instead `PIT_StopTimer`
- K82F - bug, `PIT_StartTimer` was called by `ticker free()` instead `PIT_StopTimer`
- KW24D - bug, `PIT_StartTimer` was called by `ticker free()` instead `PIT_StopTimer`
- KW41Z - bug, `PIT_StartTimer` was called by `ticker free()` instead `PIT_StopTimer`
- LPC546XX - ticker init flag has not been reset by `ticker free()`.
- NUCLEO family  - add `HAL_TIM_OC_Start(&TimMasterHandle, TIM_CHANNEL_1);` to `us_ticker_init()`.

All fixed. @0xc0170  Can we try again with morph?

cmonr commented:
@mprse Will do once the 5.9.4 RC has passed and been merged.

cmonr commented:
/morph build

0xc0170 commented:
One last  target to be fixed

mprse commented:
> Test : FAILURE
> Build number : 2424
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/7508/2424

There is still problem with `NUCLEO_F746ZG`. `tests-mbed_hal-common_tickers` has failed on all compilers (`Testcase: lp ticker interrupt test`). This is strange since this PR do not touch this test case. What differs this board from the other STM boards connected to CI is that lp ticker is based on `LPTIM` (not `RTC`). I'll try to investigate that, but I do not have this board.

@bcostm Do you maybe have some thought about this?

0xc0170 commented:
Looking at the previous test runs, this is the new issue, and seems not related here (could be an issue on master)

Restarting to confirm

/morph test

bcostm commented:
> What differs this board from the other STM boards connected to CI is that lp ticker is based on LPTIM (not RTC).

Yes you're right. I think also there should be a problem somewhere in the code managing the LPTIM. But no idea at first glance...

>  I'll try to investigate that, but I do not have this board.

This is the list of the boards having the LPTICKER/LPTIM configuration in targets.json:

DISCO_F413ZH
DISCO_F746NG
DISCO_F769NI
DISCO_L072CZ_LRWAN1
DISCO_L475VG_IOT01A
DISCO_L476VG
DISCO_L496AG
NUCLEO_F410RB
NUCLEO_F413ZH
NUCLEO_F746ZG
NUCLEO_F756ZG
NUCLEO_F767ZI
NUCLEO_L073RZ
NUCLEO_L432KC
NUCLEO_L433RC_P
NUCLEO_L476RG
NUCLEO_L486RG
NUCLEO_L496ZG
NUCLEO_L496ZG_P

Do you have one of them ?

cmonr commented:
@bcostm From that list, we only have NUCLEO_F746ZG in CI.

cmonr commented:
Wow, quite the apt timing with that last failure.

bcostm commented:
> From that list, we only have NUCLEO_F746ZG in CI.

I was just thinking that maybe you have one of the other boards on your desk/office ? I'll try to run this test on other boards.

mprse commented:
@cmonr Thanks for clarification. I have successfully reproduced the failure on NUCLEO_L476RG and I'm looking into it.

mprse committed:
tests-mbed_hal-common_tickers: use low level ticker API for time measuring.

In `ticker speed` test case execution time of ticker API functions is measured using Timer object. Test replaces original ticker handler for testing purposes, so test should not relay on higher level ticker based features(like Timer).
Use low level ticker API for time measuring.

mprse commented:
I have modified STM `lp_ticker_free()` function for boards with `LPTIM`. In current version it only disables lp ticker interrupt (does not stop the counter). This is still consistent with the requirements. 
Additionally modified `ticker speed` test case. Execution time of each ticker API function is measured using Timer object. Test replaces original ticker handler for testing purposes, so test should not relay on higher level ticker based features(like Timer) since time tracing is disrupted. Used low level ticker API for time measuring. 

@0xc0170 @cmonr Can we try again with morph?

0xc0170 commented:
/morph build

mprse commented:
`[Error] main.cpp@443,0:  #20: identifier "get_lp_ticker_data" is undefined`

Added fix for boards without lp ticker.

0xc0170 commented:
/morph build
